### PR TITLE
Added error handling

### PR DESCRIPTION
**Pull request by delvedor**
As titled.
This is a breaking change in the api.

Related: https://github.com/fastify/light-my-request/issues/10 and https://github.com/fastify/fastify/pull/617#issuecomment-355526614

~~@mcollina [Those](https://github.com/fastify/light-my-request/blob/handle-errors/test/test.js#L608-L650) test are failing, any idea why?~~
  
cc @fastify/fastify

### CODE REVIEW

File: lib/response.js [hunk 1]
```diff
@@ -20,7 +20,27 @@ function Response (req, onEnd) {
   this.once('finish', () => {
     const res = generatePayload(this)
     res.raw.req = req
-    process.nextTick(() => onEnd(res))
+    if (typeof reject === 'function') {
```
Comment (jsumners) [inline lib/response.js]: If you're doing `new Promise((resolve, reject) => {})` when would `reject` ever _not_ be a function?

Comment (delvedor) [reply lib/response.js]: In the case you are using callback style :)
https://github.com/fastify/light-my-request/blob/aeb6c202b12262f65a692d3d80a458561e4aa7b6/index.js#L67-L71

Comment (jsumners) [reply lib/response.js]: Oh, I missed that first `if` case. 👍

**Comment by mcollina** [comment]
They are failing because you should reply. Or call `destroy()` on the res connection.

**Comment by delvedor** [comment]
@mcollina but why before this pr that tests were passing?
Here I've only added an event listener.

**Comment by delvedor** [comment]
@mcollina you're right. I've updated the test. Thanks!

File: lib/response.js [hunk 2]
```diff
@@ -20,7 +20,27 @@ function Response (req, onEnd) {
...
+      process.nextTick(() => onEnd(res))
+    } else {
+      process.nextTick(() => onEnd(null, res))
+    }
+  })
+
+  this.connection.once('error', err => {
```
Comment (StarpTech) [inline lib/response.js]: Connection issues should be covered by a test too.

File: lib/response.js [hunk 3]
```diff
@@ -20,7 +20,27 @@ function Response (req, onEnd) {
...
+    if (typeof reject === 'function') {
+      process.nextTick(() => reject(err, null))
```
Comment (StarpTech) [inline lib/response.js]: Need to pass `null` ?

File: lib/response.js (hunk 2 above)
Comment (delvedor) [reply lib/response.js]: Connection errors are covered by
https://github.com/fastify/light-my-request/blob/ad9cc6e87df1235e92ffbd8ec38853245195386c/test/test.js#L757-L777

The res error has not test at the moment.

File: index.js [hunk 4]
```diff
@@ -70,9 +70,9 @@ function inject (dispatchFunc, options, callback) {
 
     return req.prepare(() => dispatchFunc.call(server, req, res))
   } else {
-    return new Promise((resolve) => {
+    return new Promise((resolve, reject) => {
       const req = new Request(options)
-      const res = new Response(req, resolve)
+      const res = new Response(req, resolve, reject)
```
Comment (nwoltman) [inline index.js]: I'd like to suggest doing something like this:

```js
const res = new Response(req, (err, res) => {
  if (err) reject(err)
  else resolve(res)
})
```

That way you wouldn't need the `if (this._promiseCallback)` checks in `lib/response.js` so `onEnd` could always just be a regular callback.

Up to you though :)

Comment (delvedor) [reply index.js]: Pr are welcome :)

**Comment by StarpTech** [comment]
Why do you pass `null` to any reject function?

**Comment by delvedor** [comment]
@StarpTech where?

**Comment by StarpTech** [comment]
@delvedor nevermind this was already fixed .. browser cache or so :D

File: lib/response.js [hunk 5]
```diff
@@ -2,10 +2,10 @@
 
 const Buffer = require('safe-buffer').Buffer
 const http = require('http')
-const Stream = require('stream')
+const stream = require('readable-stream')
```
Comment (allevo) [inline lib/response.js]: Why this change?

Comment (delvedor) [reply lib/response.js]: Compatibility down to node v4.
Socket.destroy works natively only from node v8.

Comment (allevo) [reply lib/response.js]: As always, thanks!
